---
title: "Clustering with SC3"
author: "Jordi Camps"
date: "`r format(Sys.time(), '%d %B, %Y, %H:%M')`"
output: html_document
---

# Load libraries
```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
library(knitr)
library(SC3)
library(scater)
library(RColorBrewer)
library(viridis)
library(SingleCellExperiment)
library(grid)
library(BiocGenerics)
#library(xlsx)
library(scTools)
library(dplyr)
library(lazyeval)
library(tidyr)
options(stringsAsFactors = FALSE)
```

# Import dataset
```{r}
set.seed(1234567)
sce <- readRDS("data/1moWTB_scran.rds")
endog_genes <- !rowData(sce)$is_feature_control
sce <- sce[endog_genes, ]
#sce <- sce[endog_genes, ]
#rowData(camps)$feature_symbol <- rownames(camps)
#camps <- camps[!duplicated(rownames(camps)), ]
```


# Clustering
## SC3
```{r, message = FALSE, warning=FALSE}
sce <- sc3_prepare(sce)
sce <- sc3_estimate_k(sce)
metadata(sce)$sc3$k_estimation
sce <- sc3(sce, ks = 2:11, biology = TRUE, n_cores = 1)
```

### SC3 interactive
Check dataset interactively online
```{r eval=FALSE, include=FALSE}
sc3_interactive(camps)
```

### Export results to excel
**Install perl on your computer to use this function!**
```{r}
sc3_export_results_xls(sce)
```

## Check principal components to use
```{r}
variables <- c("total_features", "total_counts", "cell_type", "genotype")
for (v in variables) {
  print(
    plotQC(sce, type = "find-pcs", variable = v) +
      ggtitle(v)
  )
}
```

```{r}
k = c(2:11)
clusters = c()
for (i in k) {
  clusters[i-1] <- paste("sc3", i, "clusters", sep = "_")
}
```


```{r fig.height=10, fig.width=10}
variables <- c("cell_type", "genotype")
var_clust <- c(variables, clusters)
for (v in var_clust) {
  print(
    plotPCA(sce, ncomponents = 10, colour_by = v) +
    ggtitle(v)
  )
}
```


### Plot clusters on PC1-7
Create dataframe of principal components to use. 
Select principal components that are important for you in **pcs dataframe**
```{r}
pcs <- c()
for (i in 1:7) {
  pcs[i] <- paste("PC", i, sep = "")
}
```

```{r}
pcs <- data.frame(x = c("PC1", "PC1", "PC1", "PC2", "PC2", "PC2", "PC2", "PC3", "PC3", "PC3", "PC3", "PC4", "PC4", "PC4", "PC5", "PC6"), 
                  y = c("PC2", "PC3", "PC4", "PC3", "PC4", "PC5", "PC6", "PC4", "PC5", "PC6", "PC7", "PC5", "PC6", "PC7", "PC7", "PC7"), 
                  stringsAsFactors = FALSE)
```

Extract components out of singlecellexperiment object
```{r}
sce=plotPCA(sce, exprs_values = "logcounts", ntop=500, colour_by = "cell_type",  ncomponents=7, theme_size =10, return_SCE = TRUE)
sce$PC1 <- reducedDim(sce)[,1]
sce$PC2 <- reducedDim(sce)[,2]
sce$PC3 <- reducedDim(sce)[,3]
sce$PC4 <- reducedDim(sce)[,4]
sce$PC5 <- reducedDim(sce)[,5]
sce$PC6 <- reducedDim(sce)[,6]
sce$PC7 <- reducedDim(sce)[,7]
```


Plot clusters over all selected PCs
```{r}
for (k in clusters) {
  for (i in 1:nrow(pcs)) {
  #sce[[k]] <- factor(sce[[k]])
  plot_components(sce, pcs[i, 1], pcs[i, 2], group = k, palette = 3, folder = "plots", subfolder = "sc3", width = 14, height = 12)
  }
}
```

### Differentially expressed genes
Export all differentially expressed genes to sce_object
```{r}
sc3_plot_de_genes(sce, k = 9, show_pdata = TRUE)
```

Export excel with significant DE genes
```{r}
sce %>%
  rowData() %>%
  as_tibble() %>%
  filter(sc3_9_de_padj < 0.01) %>%
  select(mgi_symbol, ensembl_gene_id, sc3_9_markers_auroc, sc3_9_markers_clusts, sc3_9_de_padj) %>%
  arrange(sc3_9_de_padj) %>%
  write.csv("tables/1moWTB_k9_de_genes.csv")
```


Plot DE genes on PCA
```{r}
de_k5 <- de_genes$mgi_symbol

for(g in de_k5) {
  if(g %in% row.names(camps)){
      for (i in 1:6) {
      plot_components(camps, pcs[i, 1], pcs[i, 2], group = g, folder = "sc3", subfolder = "de_k5", gene = TRUE)
    }
  }
}
```

Plot heatmap of DE genes with SC3
```{r}
library(grid)
plot_list = list()
for (x in c(2:6)) {
p <- sc3_plot_de_genes(camps, k = x, show_pdata = c("cell_type", "genotype", "total_features"))
plot_list[[x-1]] <- p
}
for (i in 1:5) {
  save_pheatmap_tiff(plot_list[[i]], paste("SC3/DE_k=",i+1,".tiff", sep = ""), 20, 20)
}
```

### Markers
Export marker genes to sce_object
```{r}
sc3_plot_markers(camps, k = 5, show_pdata = )
```


Get excel with significant marker genes
```{r}
markers <- rowData(camps)
#markers <- order(markers$sc3_5_markers_padj, na.last = TRUE)
markers <- markers[order(markers$sc3_5_markers_padj, na.last = TRUE),]
markers <- markers[!is.na(markers$sc3_5_markers_padj), ]
markers <- markers[markers$sc3_5_markers_padj < 0.01, ]
markers <- markers[, c("mgi_symbol", "sc3_5_markers_auroc", "sc3_5_markers_clusts", "sc3_5_markers_padj")]
write.xlsx(markers, "6moWTB_marker_genes.xlsx")
```

Plot marker genes with SC3
```{r, fig.height=10, fig.width=15}
options(error=recover)
ks <- 2:6
plot_list = list()
for (i in seq_along(ks)) {
p <- sc3_plot_markers(camps, k = ks[i], show_pdata = c("cell_type", "genotype", "total_features"))
plot_list[[i]] <- p
}
for (i in 1:5) {
  save_pheatmap_tiff(plot_list[[i]], paste("SC3/markers_k=",i+1,".tiff", sep = ""), width = 20, height = 20)
}
```

plot adipogenesis genes enrichr
```{r}
adipo <- c("Ncor2", "Gdf10", "Stat5b", "Socs1", "Nrip1", "Il6st", "Nr3c1")
for(g in adipo) {
  if(g %in% row.names(sce)){
      plot_components(sce, "PC2", "PC3", group = g, folder = "plots", subfolder = "adipo", gene = TRUE)
  }
}
```

```{r}
adipo_fap <- c("Gdf10", "Pdgfra", "Meox2", "Osr1")
for(g in adipo_fap) {
  if(g %in% row.names(sce)){
      plot_components(sce, "PC2", "PC3", group = g, folder = "plots", subfolder = "adipo_fap", gene = TRUE)
  }
}
```


## Save RDS
```{r}
saveRDS(camps, file = "Data/6moWTB_reads_qc_scran_SC3.rds")
```
