---
title: "Clustering with SC3"
author: "Jordi Camps"
date: "`r format(Sys.time(), '%d %B, %Y, %H:%M')`"
output: html_document
---

# Load libraries
```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
library(knitr)
library(SC3)
library(scater)
library(RColorBrewer)
library(viridis)
library(SingleCellExperiment)
library(grid)
library(BiocGenerics)
library(scTools)
library(dplyr)
library(lazyeval)
library(tidyr)
library(rlang)
options(stringsAsFactors = FALSE)
```

# Import dataset
```{r}
set.seed(1234567)
sce <- readRDS("data/1moWTB_scran.rds")
endog_genes <- !rowData(sce)$is_feature_control
sce <- sce[endog_genes, ]
#sce <- sce[endog_genes, ]
#rowData(camps)$feature_symbol <- rownames(camps)
#camps <- camps[!duplicated(rownames(camps)), ]
```


# Clustering
## SC3
```{r, message = FALSE, warning=FALSE}
sce <- sc3_prepare(sce)
sce <- sc3_estimate_k(sce)
metadata(sce)$sc3$k_estimation
sce <- sc3(sce, ks = 2:11, biology = TRUE, n_cores = 1)
```

### SC3 interactive
Check dataset interactively online
```{r eval=FALSE, include=FALSE}
sc3_interactive(camps)
```

### Export results to excel
**Install perl on your computer to use this function!**
```{r}
sc3_export_results_xls(sce)
```

## Check principal components to use
```{r}
variables <- c("total_features", "total_counts", "cell_type", "genotype")
for (v in variables) {
  print(
    plotQC(sce, type = "find-pcs", variable = v) +
      ggtitle(v)
  )
}
```


```{r}
k = c(2:11)
clusters = c()
for (i in k) {
  clusters[i-1] <- paste("sc3", i, "clusters", sep = "_")
}
```

Show heatmap of all selected ks 
```{r fig.height=12, fig.width=10}
for (k in 2:11) {
    sc3_plot_markers(sce, k = k, auroc = 0.9, show_pdata = c("cell_type", "genotype", "total_features")) 
} 
```
ks <- 2:6
plot_list = list()
for (i in seq_along(ks)) {
p <- sc3_plot_markers(camps, k = ks[i], show_pdata = c("cell_type", "genotype", "total_features"))
plot_list[[i]] <- p
}

```{r fig.height=10, fig.width=10}
variables <- c("cell_type", "genotype")
var_clust <- c(variables, clusters)
for (v in var_clust) {
  print(
    plotPCA(sce, ncomponents = 10, colour_by = v) +
    ggtitle(v)
  )
}
```


### Plot clusters on PC1-7
Create dataframe of principal components to use. 
Select principal components that are important for you in **pcs dataframe**
```{r}
pcs <- c()
for (i in 1:7) {
  pcs[i] <- paste("PC", i, sep = "")
}
```

```{r}
pcs <- data.frame(x = c("PC1", "PC1", "PC1", "PC2", "PC2", "PC2", "PC2", "PC3", "PC3", "PC3", "PC3", "PC4", "PC4", "PC4", "PC5", "PC6"), 
                  y = c("PC2", "PC3", "PC4", "PC3", "PC4", "PC5", "PC6", "PC4", "PC5", "PC6", "PC7", "PC5", "PC6", "PC7", "PC7", "PC7"), 
                  stringsAsFactors = FALSE)
```

Extract components out of singlecellexperiment object
```{r}
sce=plotPCA(sce, exprs_values = "logcounts", ntop=500, colour_by = "cell_type",  ncomponents=7, theme_size =10, return_SCE = TRUE)
sce$PC1 <- reducedDim(sce)[,1]
sce$PC2 <- reducedDim(sce)[,2]
sce$PC3 <- reducedDim(sce)[,3]
sce$PC4 <- reducedDim(sce)[,4]
sce$PC5 <- reducedDim(sce)[,5]
sce$PC6 <- reducedDim(sce)[,6]
sce$PC7 <- reducedDim(sce)[,7]
```


Plot clusters over all selected PCs
```{r}
for (k in clusters) {
  for (i in 1:nrow(pcs)) {
  #sce[[k]] <- factor(sce[[k]])
  plot_components(sce, pcs[i, 1], pcs[i, 2], group = k, palette = "Set2", folder = "plots", subfolder = "sc3", width = 14, height = 12)
  }
}
```

###PC1 and 5 dominated by total genes
```{r fig.height=6, fig.width=6}
temp <- data.frame(PC1 = sce[["PC1"]], total_features = sce[["total_features"]], col = sce[["cell_type"]])
ggplot(temp, aes(x = total_features, y = PC1)) +
         geom_point(size = 3, aes(col = col)) +
         geom_smooth(method = "lm", colour = "grey20") +
         labs(y = "Principal component 1", x = "Total detected genes", color = "FACS marker") +
         scale_color_brewer(type = "qual", palette = 2) +
         theme_bw(base_size = 18) +
         theme(panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
legend.title = element_blank()) +
  ggsave("plots/PCA/PC1_total_features_cell_type.tiff", width = 14, height = 12, units = "cm") 
       
```


### Smooth muscle cells(k=6, cluster 5 and 6)
```{r}
clusters <- c("sc3_4_clusters", "sc3_6_clusters")
pcs <- data.frame(x = c("PC2", "PC3", "PC2", "PC3"),
                  y = c("PC3", "PC4", "PC4", "PC6"),
                  stringsAsFactors = FALSE)
for (k in clusters) {
  for (i in 1:nrow(pcs)) {
plot_components(sce, pcs[i, 1], pcs[i, 2], group = k, save = TRUE, folder = "plots", subfolder = "SMC")
  }
}
```

Plot genes specific to cluster 5 and 6 (k=6)
```{r}
select_markers(sce, k = 6, clust_num = 5, vector_name = "genes")
pcs <- data.frame(x = c("PC2", "PC3", "PC2", "PC3"),
                  y = c("PC3", "PC4", "PC4", "PC6"),
                  stringsAsFactors = FALSE)

for(g in genes) {
  if(g %in% row.names(sce)){
    for (i in 1:nrow(pcs)) {
      plot_components(sce, pcs[i, 1], pcs[i, 2], group = g, folder = "plots", subfolder = "SMC", alpha = 0.8, gene = TRUE)
    }
  }
}
```

```{r}
select_markers(sce, k = 6, clust_num = 6, vector_name = "genes")
pcs <- data.frame(x = c("PC2", "PC3", "PC2", "PC3"),
                  y = c("PC3", "PC4", "PC4", "PC6"),
                  stringsAsFactors = FALSE)

for(g in genes) {
  if(g %in% row.names(sce)){
    for (i in 1:nrow(pcs)) {
      plot_components(sce, pcs[i, 1], pcs[i, 2], group = g, folder = "plots", subfolder = "SMC", alpha = 0.8, gene = TRUE)
    }
  }
}
```

Plot genes specific to cluster 4 (k=4)
```{r}
select_markers(sce, k = 4, clust_num = 4, vector_name = "genes")

pcs <- data.frame(x = c("PC2", "PC3", "PC2", "PC3"),
                  y = c("PC3", "PC4", "PC4", "PC6"),
                  stringsAsFactors = FALSE)

for(g in genes) {
  if(g %in% row.names(sce)){
    for (i in 1:nrow(pcs)) {
      plot_components(sce, pcs[i, 1], pcs[i, 2], group = g, folder = "plots", subfolder = "SMC", alpha = 0.8, gene = TRUE)
    }
  }
}
```

### Satellite cells
plot genes cluster 3 9k = 4)
```{r}
select_markers(sce, k = 4, clust_num = 3, vector_name = "genes")

pcs <- data.frame(x = c("PC2", "PC3", "PC2", "PC3"),
                  y = c("PC3", "PC4", "PC4", "PC6"),
                  stringsAsFactors = FALSE)

#genes <- c("Myf5", "Fgfr4")

for(g in genes) {
  if(g %in% row.names(sce)){
    for (i in 1:nrow(pcs)) {
      plot_components(sce, pcs[i, 1], pcs[i, 2], group = g, folder = "plots", subfolder = "muSC", alpha = 0.8, gene = TRUE)
    }
  }
}
```

### Component 1 and 3
```{r}
clusters <- c("sc3_4_clusters", "sc3_6_clusters")
#pcs <- data.frame(x = c("PC1", "PC3", "PC2", "PC3"),
#                  y = c("PC3", "PC4", "PC4", "PC6"),
#                  stringsAsFactors = FALSE)
for (k in clusters) {
 # for (i in 1:nrow(pcs)) {
plot_components(sce, "PC1", "PC3", group = k, save = TRUE, folder = "plots", subfolder = "cell_type", palette = 2)
  }
#}
```


```{r}
for (i in 1:4) {
  select_markers(sce, k = 4, clust_num = i, vector_name = paste0("genes_cluster_", i, sep = ""))
}

genes_cluster <- list(c(genes_cluster_1, genes_cluster_2, genes_cluster_3, genes_cluster_4))


for(g in genes_cluster_1) {
    if(g %in% row.names(sce)){
      #for (i in 1:nrow(pcs)) {
        plot_components(sce, "PC1", "PC3", group = g, folder = "plots", subfolder = "k4_cluster_1", alpha = 0.8, gene = TRUE)
    #}
  }
}

for(g in genes_cluster_2) {
    if(g %in% row.names(sce)){
      #for (i in 1:nrow(pcs)) {
        plot_components(sce, "PC1", "PC3", group = g, folder = "plots", subfolder = "k4_cluster_2", alpha = 0.8, gene = TRUE)
    #}
  }
}

for(g in genes_cluster_3) {
    if(g %in% row.names(sce)){
      #for (i in 1:nrow(pcs)) {
        plot_components(sce, "PC1", "PC3", group = g, folder = "plots", subfolder = "k4_cluster_3", alpha = 0.8, gene = TRUE)
    #}
  }
}

for(g in genes_cluster_4) {
    if(g %in% row.names(sce)){
      #for (i in 1:nrow(pcs)) {
        plot_components(sce, "PC1", "PC3", group = g, folder = "plots", subfolder = "k4_cluster_4", alpha = 0.8, gene = TRUE)
    #}
  }
}
```

### Cluster 5
k = 8, cluster 5 FMOD cells
```{r}
plot_components(sce, "PC4", "PC7", group = "sc3_8_clusters", save = TRUE, folder = "plots", subfolder = "FMOD_cells", palette = 2)
```

```{r}
select_markers(sce, k = 8, clust_num = 5)

for(g in genes) {
    if(g %in% row.names(sce)){
      #for (i in 1:nrow(pcs)) {
        plot_components(sce, "PC4", "PC7", group = g, folder = "plots", subfolder = "k8_cluster_5", alpha = 0.8, gene = TRUE)
    #}
  }
}
```


### Genotype components
Check PC6 and 7 (genotype components) in PC2-5 (cell type components)
```{r}
clusters <- c("sc3_4_clusters", "sc3_8_clusters", "sc3_9_clusters")

x <- c(rep("PC6", 4), rep("PC7", 5))
y <- c(rep(paste0("PC", 2:5), 2), "PC6")
pcs <- data.frame(x = x,
                  y = y,
                  stringsAsFactors = FALSE)

for (k in clusters) {
  for (i in 1:nrow(pcs)) {
plot_components(sce, pcs[i, 1], pcs[i, 2], group = k, save = TRUE, palette = 2, folder = "plots", subfolder = "genotype")
  }
}

#Own colors for k = 4
#for (k in clusters) {
  for (i in 1:nrow(pcs)) {
plot_components(sce, pcs[i, 1], pcs[i, 2], group = "sc3_4_clusters", save = TRUE, brewer = FALSE, hex_codes = c("#1b9e77", "#d95f02", "#7570b3", "#e7298a"), folder = "plots", subfolder = "genotype_fap")
  }
#}

#Own colors for k = 8
  for (i in 1:nrow(pcs)) {
plot_components(sce, pcs[i, 1], pcs[i, 2], group = "sc3_8_clusters", save = TRUE, brewer = FALSE, hex_codes = c("#1b9e77", "#d95f02",  rep("#999999", 6)), folder = "plots", subfolder = "genotype_fap")
  }

#OWn colors for k = 9
  for (i in 1:nrow(pcs)) {
plot_components(sce, pcs[i, 1], pcs[i, 2], group = "sc3_9_clusters", save = TRUE, brewer = FALSE, hex_codes = c("#1b9e77", "#d95f02", "#7570b3", rep("#999999", 6)), folder = "plots", subfolder = "genotype_fap")
  }

#for (k in clusters) {
  for (i in 1:nrow(pcs)) {
plot_components(sce, pcs[i, 1], pcs[i, 2], group = "genotype", save = TRUE, brewer = FALSE, hex_codes = c("#FFC90C", "#89CAFF"), folder = "plots", subfolder = "genotype")
  }
#}
```

#### Sca1+ cells
Markers for k = 8
```{r}
for (i in 1:2) {
  select_markers(sce, k = 8, clust_num = i, vector_name = paste0("genes_cluster_", i, sep = ""))
}

for(g in genes_cluster_1) {
    if(g %in% row.names(sce)){
      for (i in 1:nrow(pcs)) {
        plot_components(sce, pcs[i, 1], pcs[i, 2], group = g, folder = "plots", subfolder = "k8_cluster_1", alpha = 0.8, gene = TRUE)
    }
  }
}

for(g in genes_cluster_2) {
    if(g %in% row.names(sce)){
      for (i in 1:nrow(pcs)) {
        plot_components(sce, pcs[i, 1], pcs[i, 2], group = g, folder = "plots", subfolder = "k8_cluster_2", alpha = 0.8, gene = TRUE)
    }
  }
}
```


markers for k = 9
```{r}
for (i in 1:3) {
  select_markers(sce, k = 9, clust_num = i, vector_name = paste0("genes_cluster_", i, sep = ""))
}

for(g in genes_cluster_1) {
    if(g %in% row.names(sce)){
      for (i in 1:nrow(pcs)) {
        plot_components(sce, pcs[i, 1], pcs[i, 2], group = g, folder = "plots", subfolder = "k9_cluster_1", alpha = 0.8, gene = TRUE)
    }
  }
}

for(g in genes_cluster_2) {
    if(g %in% row.names(sce)){
      for (i in 1:nrow(pcs)) {
        plot_components(sce, pcs[i, 1], pcs[i, 2], group = g, folder = "plots", subfolder = "k9_cluster_2", alpha = 0.8, gene = TRUE)
    }
  }
}

for(g in genes_cluster_3) {
    if(g %in% row.names(sce)){
      for (i in 1:nrow(pcs)) {
        plot_components(sce, pcs[i, 1], pcs[i, 2], group = g, folder = "plots", subfolder = "k9_cluster_3", alpha = 0.8, gene = TRUE)
    }
  }
}


    if("Cd55" %in% row.names(sce)){
      for (i in 1:nrow(pcs)) {
        plot_components(sce, pcs[i, 1], pcs[i, 2], group = "Cd55", folder = "plots", subfolder = "k9_cluster_3", alpha = 0.8, gene = TRUE)
    }
  }
```

#### muSCs
genotype differences in component 3 and 7. Although these cells don't cluster differently!? Look at gene weights to check which differences are there between these cells.

Export gene weights to csv file
```{r}
rv <- rowVars(assay(sce))
selected_genes <- order(rv, decreasing = TRUE)[seq_len(min(500, length(rv)))]
pca <- prcomp(t(assay(sce)[selected_genes,]))
pca_loadings <- as.data.frame(pca$rotation)
pca_loadings[order(pca_loadings[,1], decreasing = TRUE),]
write.csv(pca_loadings, "tables/PCA_gene_weights.csv")
```

Plot genes with high weights in PC7
```{r}
genes <- c("Fos", "Sparc", "Apoe", "X.9", "Col3a1", "Lyz2", "Apod", "Pmp22", "Matn2", "Gsn", "Malat1", "Sat1", "Tinagl1", "Postn", "Gpc3", "Hspa1a", "Gas7", "Dbi", "Gpm6b", "H2.D1", "Cd9", "Gfra3", "Cst3", "Jun", "Hspa5", "Pdlim4", "Gpx3", "Myf5", "Myog", "Myod1", "Pax7", "Mymk")

for(g in genes) {
    if(g %in% row.names(sce)){
      #for (i in 1:nrow(pcs)) {
        plot_components(sce, "PC7", "PC3", group = g, folder = "plots", subfolder = "genotype_muSC", alpha = 0.8, gene = TRUE)
    #}
  }
}
```

Get ensembl ID of X.9 genes
Seems to be ribosomal RNA
```{r}
X.9 <- sce["X.9",]
rowData(X.9)
```


### Differentially expressed genes
Export all differentially expressed genes to sce_object

Function to filter for genes in clusters, padj and auroc
export all significant genes above auroc to csv file of a designated k
```{r}
export_markers(sce, k = 6, folder = "tables", filename = "1moWTB_k6_markers")
```

Select significant markers
```{r}
select_markers(sce, k = 6, clust_num = 1, vector_name = "genes")
```


#### k=6
```{r}
sc3_plot_de_genes(sce, k = 6, show_pdata = TRUE)
```

Export excel with significant DE genes
```{r}
sce %>%
  rowData() %>%
  as_tibble() %>%
  filter(sc3_6_de_padj < 0.01) %>%
  select(mgi_symbol, ensembl_gene_id, sc3_6_markers_auroc, sc3_6_markers_clusts, sc3_6_de_padj) %>%
  arrange(sc3_6_de_padj) %>%
  write.csv("tables/1moWTB_k6_de_genes.csv")
```



#### k=9
```{r}
sc3_plot_de_genes(sce, k = 9, show_pdata = TRUE)
```

Export excel with significant DE genes
```{r}
sce %>%
  rowData() %>%
  as_tibble() %>%
  filter(sc3_9_de_padj < 0.01) %>%
  select(mgi_symbol, ensembl_gene_id, sc3_9_markers_auroc, sc3_9_markers_clusts, sc3_9_de_padj) %>%
  arrange(sc3_9_de_padj) %>%
  write.csv("tables/1moWTB_k9_de_genes.csv")
```


Plot DE genes on PCA
```{r}
de_k5 <- de_genes$mgi_symbol

for(g in de_k5) {
  if(g %in% row.names(camps)){
      for (i in 1:6) {
      plot_components(camps, pcs[i, 1], pcs[i, 2], group = g, folder = "sc3", subfolder = "de_k5", gene = TRUE)
    }
  }
}
```

Plot heatmap of DE genes with SC3
```{r}
library(grid)
plot_list = list()
for (x in c(2:6)) {
p <- sc3_plot_de_genes(camps, k = x, show_pdata = c("cell_type", "genotype", "total_features"))
plot_list[[x-1]] <- p
}
for (i in 1:5) {
  save_pheatmap_tiff(plot_list[[i]], paste("SC3/DE_k=",i+1,".tiff", sep = ""), 20, 20)
}
```

### Markers
#### k = 4
```{r}
sc3_plot_markers(sce, k = 4, show_pdata = TRUE)
```

```{r}
sce %>%
  rowData() %>%
  as_tibble() %>%
  group_by(sc3_4_markers_clusts) %>%
  filter(sc3_4_markers_padj < 0.01 & sc3_4_markers_auroc > 0.8) %>%
  select(mgi_symbol, ensembl_gene_id, sc3_4_markers_auroc, sc3_4_markers_padj) %>%
  arrange(sc3_4_markers_padj, .by_group = TRUE) %>%
  write.csv("tables/1moWTB_k4_markers.csv")
```

#### k = 6
Export marker genes to sce_object
```{r}
sc3_plot_markers(sce, k = 6, show_pdata = TRUE)
```


Get excel with significant marker genes
```{r}
sce %>%
  rowData() %>%
  as_tibble() %>%
  group_by(sc3_6_markers_clusts) %>%
  filter(sc3_6_markers_padj < 0.01 & sc3_6_markers_auroc > 0.8) %>%
  select(mgi_symbol, ensembl_gene_id, sc3_6_markers_auroc, sc3_6_markers_padj) %>%
  arrange(sc3_6_markers_padj, .by_group = TRUE) %>%
  write.csv("tables/1moWTB_k6_markers.csv")
```



Plot marker genes with SC3
```{r, fig.height=10, fig.width=15}
options(error=recover)
ks <- 2:6
plot_list = list()
for (i in seq_along(ks)) {
p <- sc3_plot_markers(camps, k = ks[i], show_pdata = c("cell_type", "genotype", "total_features"))
plot_list[[i]] <- p
}
for (i in 1:5) {
  save_pheatmap_tiff(plot_list[[i]], paste("SC3/markers_k=",i+1,".tiff", sep = ""), width = 20, height = 20)
}
```

plot adipogenesis genes enrichr
```{r}
adipo <- c("Ncor2", "Gdf10", "Stat5b", "Socs1", "Nrip1", "Il6st", "Nr3c1")
for(g in adipo) {
  if(g %in% row.names(sce)){
      plot_components(sce, "PC2", "PC3", group = g, folder = "plots", subfolder = "adipo", gene = TRUE)
  }
}
```

```{r}
adipo_fap <- c("Gdf10", "Pdgfra", "Meox2", "Osr1")

for(g in adipo_fap) {
  if(g %in% row.names(sce)){
      plot_components(sce, "PC2", "PC3", group = g, folder = "plots", subfolder = "adipo_fap", gene = TRUE)
  }
}
```

```{r}
fibro_fap <- c("Igf1")

for(g in fibro_fap) {
  if(g %in% row.names(sce)){
      plot_components(sce, "PC2", "PC3", group = g, folder = "plots", subfolder = "fibro_fap", gene = TRUE)
  }
}
```


## Save RDS
```{r}
saveRDS(camps, file = "Data/6moWTB_reads_qc_scran_SC3.rds")
```
